Math/NumberTheory/GaussianIntegers.hs

```diff
@@ -0,0 +1,233 @@
+divG :: GaussianInteger -> GaussianInteger -> GaussianInteger
+n `divG` d = q where (q,_) = divModG n d
+
+-- |Gaussian integer remainder, satisfying
+--
+-- > (x `divG` y)*y + (x `modG` y) == x
+modG :: GaussianInteger -> GaussianInteger -> GaussianInteger
+n `modG` d = r where (_,r) = divModG n d
+
+divHelper :: (Integer -> Integer -> Integer) -> GaussianInteger -> GaussianInteger -> (GaussianInteger, GaussianInteger)
+divHelper divide g h =
+    let nr :+ ni = g * conjugate h
+        denom = norm h
+        q = divide nr denom :+ divide ni denom
+        p = h * q
+    in (q, g - p)
+
+-- |Conjugate a Gaussian integer.
+conjugate :: GaussianInteger -> GaussianInteger
+conjugate (r :+ i) = r :+ (-i)
+
+-- |The square of the magnitude of a Gaussian integer.
+norm :: GaussianInteger -> Integer
+norm (x :+ y) = x * x + y * y
+
+-- |Compute whether a given Gaussian integer is prime.
+isPrime :: GaussianInteger -> Bool
+isPrime g@(x :+ y)
+    | x == 0 && y /= 0 = abs y `mod` 4 == 3 && Testing.isPrime y
+    | y == 0 && x /= 0 = abs x `mod` 4 == 3 && Testing.isPrime x
+    | otherwise        = Testing.isPrime $ norm g
+
+-- |An infinite list of the Gaussian primes. Uses primes in Z to exhaustively
+-- generate all Gaussian primes, but not quite in order of ascending magnitude.
+primes :: [GaussianInteger]
+primes = [ a' :+ b'
+         | p <- Sieve.primes
+         , (a', b') <- if p `mod` 4 == 3
+             then [ (p, 0) ]
+             else [ (a, b)
+                  | let radius = Powers.integerSquareRoot p
+                  , a <- [0 .. radius]
+                  , let d = p - a * a
+                  , Powers.isSquare d
+                  , let b = Powers.integerSquareRoot d
+                  ]
+         ]
+
+-- |Compute the GCD of two Gaussian integers. Enforces the precondition that each
+-- integer must be in the first quadrant (or zero).
+gcdG :: GaussianInteger -> GaussianInteger -> GaussianInteger
+gcdG g h = gcdG' (abs g) (abs h)
+
+-- |Compute the GCD of two Gauss integers. Does not check the precondition.
+gcdG' :: GaussianInteger -> GaussianInteger -> GaussianInteger
+gcdG' g h
+    | h == 0    = g --done recursing
+    | otherwise = gcdG' h (abs (g `modG` h))
+
+-- |Find a Gaussian integer whose norm is the given prime number.
+-- Checks the precondition that p is prime and that p `mod` 4 /= 3.
+findPrime :: Integer -> GaussianInteger
+findPrime p
+    | p == 2 || (p `mod` 4 == 1 && Testing.isPrime p) = findPrime' p
+    | otherwise = error "p must be prime, and not congruent to 3 (mod 4)"
+
+-- |Find a Gaussian integer whose norm is the given prime number. Does not
+-- check the precondition.
+findPrime' :: Integer -> GaussianInteger
+findPrime' p =
+    let (Just c) = Moduli.sqrtModP (-1) p
+        k  = Powers.integerSquareRoot p
+        bs = [1 .. k]
+        as = map (\b' -> (b' * c) `mod` p) bs
+        a  = head [ a' | a' <- as, a' <= k]
+        b  = head [ b' | b' <- bs, b' * b' + a * a == p]
+    in a :+ b
```

Comment:
IMHO it is more idiomatic to zip `bs` and `as` together, when we need to find correspondent elements. E. g., `abs = map (\b' -> (b' * c `mod` p, b') bs`. 

I am not sure that `head` is a right choice here. If a user supply unsupported prime (of form 4k+3), he will get confusing `Exception: Prelude.head: empty list`. It would be great to show custom error message with a relevant hint.